Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift

```diff
@@ -0,0 +1,160 @@
+import Foundation
+import CardinalMobile
+
+#if canImport(BraintreeCore)
+import BraintreeCore
+#endif
+
+// TODO: Can be internal and maybe a struct once BTThreeDSecureRequest is in Swift
+@objcMembers public class BTThreeDSecureV2Provider: NSObject {
+
+    // MARK: - Internal Properties
+
+    let cardinalSession: CardinalSession
+    let apiClient: BTAPIClient
+
+    var lookupResult: BTThreeDSecureResult? = nil
+
+    static var completionHandler: (BTThreeDSecureResult?, Error?) -> Void = { _, _ in }
+
+    // MARK: - Initializer
+
+    // TODO: can be internal and non obj-c when BTThreeDSecureRequest is in Swift
+    @objc(initWithConfiguration:apiClient:request:completion:)
+    public init(
+        configuration: BTConfiguration,
+        apiClient: BTAPIClient,
+        request: BTThreeDSecureRequest,
+        completion: @escaping ([String: String]?) -> Void
+    ) {
+        self.apiClient = apiClient
+        self.cardinalSession = CardinalSession()
+
+        let cardinalConfiguration: CardinalSessionConfiguration = CardinalSessionConfiguration()
+
+        if let v2UICustomization = request.v2UICustomization {
+            cardinalConfiguration.uiCustomization = v2UICustomization.cardinalValue
+        }
+
+        var cardinalEnvironment: CardinalSessionEnvironment = .staging
+
+        if configuration.environment == "production" {
+            cardinalEnvironment = .production
+        }
+
+        guard let cardinalAuthenticationJWT = configuration.cardinalAuthenticationJWT else {
+            completion(nil)
+            return
+        }
+
+        cardinalConfiguration.deploymentEnvironment = cardinalEnvironment
+        cardinalSession.configure(cardinalConfiguration)
+        cardinalSession.setup(
+            jwtString: cardinalAuthenticationJWT,
+            completed: { consumerSessionID in
+                apiClient.sendAnalyticsEvent("ios.three-d-secure.cardinal-sdk.init.setup-completed")
+                completion(["dfReferenceId": consumerSessionID])
+            }, validated: { _ in
+                apiClient.sendAnalyticsEvent("ios.three-d-secure.cardinal-sdk.init.setup-failed")
+                completion([:])
+            }
+        )
+    }
+
+    // MARK: - Internal Methods
+
+    // TODO: can be internal when BTThreeDSecureRequest is in Swift
+    @objc(processLookupResult:completion:)
+    public func process(
+        lookupResult: BTThreeDSecureResult,
+        completion: @escaping (BTThreeDSecureResult?, Error?) -> Void
+    ) {
+        self.lookupResult = lookupResult
+        BTThreeDSecureV2Provider.completionHandler = completion
+
+        cardinalSession.continueWith(
+            transactionId: lookupResult.lookup?.transactionID ?? "",
+            payload: lookupResult.lookup?.paReq ?? "",
+            validationDelegate: self
+        )
+    }
+
+    // MARK: - Private Methods
+
+    private func callFailureHandler(
```

Comment:
```suggestion
    private func notifyError(
```
Take or leave, we've been using `notifyX` on Android to callout when we pass a message back to the merchant.